### PR TITLE
Ray Disaggregated Serving MVP

### PR DESCRIPTION
**Pull request by FanhaiLu1**
This PR enable pytorch engine disaggregated serving on multiple TPU POD slices. 

### This PR delivered:

1.  Engine do prefill in one POD slice and do decode in another POD slice 
2. Transfer prefill result from one POD slice to another POD slice 
3. Load weight and sharding in multiple host on multiple POD slice
4. Compute meaningful decode result 

### Result validation:

1. Disaggregated serving, here is the logs (tpu-vm-1 as decode engine):

> ---- Do prefill in prefill engine pod_slice_name: tpu-vm-2
> ---- Transfer prefill result to decode engine pod_slice_name: tpu-vm-1
> ---- Do insert in decode engine pod_slice_name: tpu-vm-1

2. Correct result compared with interleave serving 

Command: 
`python /home/{user}/jetstream-pytorch/run_interactive_disaggregated.py --size=7b --batch_size=1 --is_disaggregated=True --num_hosts=8 --decode_pod_slice_name={user}-tpu-vm-2 --model_name=llama-2  --max_cache_length=2048 --quantize_weights=False --quantize_kv_cache=False --checkpoint_path=/home/{user}/data/llama-2-7b-chat-safetensor/model.safetensors --tokenizer_path=/home/{user}/data/tokenizer.model --sharding_config=/home/{user}/jetstream-pytorch/default_shardings/llama.yaml`

#### Interleave result:

```

to find purpose and fulfillment.

I believe that everyone has a unique purpose and that it is up to each individual to discover and pursue theirs.

```

#### Disaggregated result:
```

to find purpose and fulfillment.

I believe that everyone has a unique purpose and that it is up to each individual to discover and pursue theirs.

```

### Next Steps:

1.  Integrate with jetstream orchestrator 
2. Add readme for both interactive run and jetstream run 
3. Fix performance issue 
4. Test llama 70 
5: Support multiple prefill engine and multiple decode engine

### CODE REVIEW

File: jetstream_pt/ray_engine.py [hunk 1]
```diff
@@ -76,6 +82,65 @@ def prefill(
...
+  def prefill(
+      self,
+      *,
+      params: Any,  # Weights
+      existing_prefix: Optional[Prefix] = None,
+      padded_tokens: np.ndarray,  # PrefillInputs[np.ndarray],
+      true_length: int,
+  ) -> Prefix:
+    result = None
+    if self.is_disaggregated:
```
Comment (wang2yn84) [inline jetstream_pt/ray_engine.py:115]: Consider use call_prefill = worker.prefill_ray_disaggregation if self.is_disaggregated else worker.prefill_ray.

File: jetstream_pt/ray_engine.py [hunk 2]
```diff
@@ -76,6 +82,65 @@ def prefill(
     # the worker itself manages and maintains the prefill states.
     return None
 
+  # pylint: disable-next=all
+  def disaggregated_prefill(
```
Comment (wang2yn84) [inline jetstream_pt/ray_engine.py:86]: Nit: this function doesn't seem quite different from interleave_prefill either, consider consolidate in the future if not now.

**Review by wang2yn84 — APPROVED** [review]

File: jetstream_pt/ray_engine.py (hunk 1 above)
Comment (FanhaiLu1) [reply jetstream_pt/ray_engine.py]: Thanks, updated.

File: jetstream_pt/ray_engine.py (hunk 2 above)
Comment (FanhaiLu1) [reply jetstream_pt/ray_engine.py:86]: Sounds good, I may have extra change on disaggregated function, will consider to merge it in future if possible.

Comment (FanhaiLu1) [reply jetstream_pt/ray_engine.py]: Changed,  I removed the interleave and disaggregated prefill function. Both interleave and disaggregated call prefill function now,

**Comment by FanhaiLu1** [comment]
> High level comment - it looks like the main difference for `is_disaggregated` within `PyTorchRayEngine` is whether or not prefill returns outputs.
> 
> If the prefill/decode/interleave functionality is essentially the same, then I guess it's an implementation detail for orchestrator to trigger the transfer. If so, then it possible to exclude `is_disaggregated` from the worker? That'd simplify the complexity

Simplified the prefill call from engine side. On the worker side. Yes, they are same on insert and decode side. But I feel it's better to keep disaggregated and interleave for prefill. Server reasons:

1. Worker don't need handle the logic of disaggregated or interleave. Worker can do both interleave and disaggregated, let the engine chose what worker can do.
2. There are large difference between disaggregated or interleave prefill. From interleave side, we directly save the prefill result in it's local HBM. But from disaggregated side, we need extract call to all gather the cache and return the prefill result. The another difference is that pre result cache is jax array in interleave, but it's np array disaggregated
3. Worker already had pretty complex logic, it's better to keep some logic in engine (engine has simple code logic right now)

File: install_everything.sh [hunk 3]
```diff
@@ -19,9 +19,17 @@ pip show libtpu-nightly && pip uninstall -y libtpu-nightly
...
+pip show keras && pip uninstall -y keras
+pip show tensorboard && pip uninstall -y tensorboard
+pip show tensorflow-text && pip uninstall -y tensorflow-text
+pip show torch_xla2 && pip uninstall -y torch_xla2
 
 pip install flax==0.8.3
 pip install jax[tpu]==0.4.28 -f https://storage.googleapis.com/jax-releases/libtpu_releases.html
+pip install tensorflow-text
+pip install tensorflow
+pip install flax==0.8.3
```
[the remark this answers is not in the document]
Comment (FanhaiLu1) [reply install_everything.sh]: Should be just once, removed the duplicated one.

**Comment by allenwang28** [comment]
> But I feel it's better to keep disaggregated and interleave for prefill. Several reasons:

I think that makes sense to me, thanks!

File: jetstream_pt/ray_engine.py [hunk 4]
```diff
@@ -153,7 +219,10 @@ def create_pytorch_ray_engine(
     quantize_kv=False,
     max_cache_length=1024,
     sharding_config=None,
-) -> PyTorchRayEngine:
+    is_disaggregated: bool = False,
+    num_hosts: int = 0,
+    decode_pod_slice_name: str = None,
+) -> Any:
```
[the remark this answers is not in the document]
Comment (FanhaiLu1) [reply jetstream_pt/ray_engine.py:181]: Create an issue: https://github.com/google/jetstream-pytorch/issues/107 and added the issue in the comment.

**Pull request merged**